spectator-ext-jvm/src/main/java17/com/netflix/spectator/jvm/JavaFlightRecorder.java

```diff
@@ -0,0 +1,107 @@
+package com.netflix.spectator.jvm;
+
+import com.netflix.spectator.api.Registry;
+import jdk.jfr.EventSettings;
+import jdk.jfr.consumer.RecordedEvent;
+import jdk.jfr.consumer.RecordingStream;
+
+import java.time.Duration;
+import java.util.Objects;
+import java.util.concurrent.Executor;
+import java.util.concurrent.ExecutorService;
+import java.util.function.Consumer;
+
+/**
+ * Helpers supporting continuous monitoring with Java Flight Recorder.
+ */
+public class JavaFlightRecorder {
+
+  private static final String PREFIX = "jdk.";
+  private static final String CPULoad = PREFIX + "CPULoad";
+  private static final String VirtualThreadPinned = PREFIX + "VirtualThreadPinned";
+  private static final String VirtualThreadSubmitFailed = PREFIX + "VirtualThreadSubmitFailed";
+  private static final String ZAllocationStall = PREFIX + "ZAllocationStall";
+  private static final String ZYoungGarbageCollection = PREFIX + "ZYoungGarbageCollection";
+  private static final String ZOldGarbageCollection = PREFIX + "ZOldGarbageCollection";
+
+  private JavaFlightRecorder() {
+  }
+
+  /**
+   * Collect low-overhead Java Flight Recorder events, using the provided
+   * {@link Executor} to execute a single task to collect events.
+   *
+   * @param registry the registry
+   * @param executor the executor to execute the task for streaming events
+   * @return an {@link AutoCloseable} allowing the underlying
+   * {@link jdk.jfr.consumer.EventStream} to be closed
+   */
+  public static AutoCloseable monitorDefaultEvents(Registry registry, Executor executor) {
+    Objects.requireNonNull(registry);
+    Objects.requireNonNull(executor);
+    RecordingStream rs = new RecordingStream();
+    collectCpuEvents(registry, rs);
+    collectVirtualThreadEvents(registry, rs);
+    collectZgcEvents(registry, rs);
+    executor.execute(rs::start);
+    return rs::close;
+  }
+
+  private static void collectCpuEvents(Registry registry, RecordingStream rs) {
+    consume(CPULoad, rs, event -> {
+      String pid = "unknown";
+      try {
+        pid = String.valueOf(ProcessHandle.current().pid());
+      } catch (UnsupportedOperationException ignore) {
+      }
+      registry.gauge("jvm.cpuLoad", "type", "jvmUser", "pid", pid)
```

Comment:
Process IDs are not appropriate for metric dimensions. They aren't meaningful outside the node and can have really high cardinality in the event the process is restarting frequently.

Reply:
I was looking for something to differentiate multiple JVM processes, but likely moot if we feel this doesn't have utility over an above the existing CPU instrumentation.

Reply:
Differentiating processes can be done by setting the [NETFLIX_PROCESS_NAME](https://netflix.github.io/atlas-docs/spectator/lang/go/migrations/#common-tags) to a meaningful value. That then applies to all metrics for that process.

Reply:
Ah nice. Okay, I'll remember that for next time, but for now I don't think there's an immediate need for these, so I've dropped them.